compiler/rustc_lint/src/builtin.rs

```diff
@@ -2420,3 +2420,37 @@ impl<'tcx> LateLintPass<'tcx> for ClashingExternDeclarations {
         }
     }
 }
+
+declare_lint! {
+    FUNCTION_REFERENCES,
+    Warn,
+    "suggest casting functions to pointers when attempting to take references"
+}
+
+declare_lint_pass!(FunctionReferences => [FUNCTION_REFERENCES]);
+
+impl<'tcx> LateLintPass<'tcx> for FunctionReferences {
+    fn check_expr(&mut self, cx: &LateContext<'_>, e: &hir::Expr<'_>) {
+        if let hir::ExprKind::AddrOf(hir::BorrowKind::Ref, _, referent) = e.kind {
```

Comment:
I do not think it is necessary to restrict this to `hir::BorrowKind::Ref` -- all `AddrOf` should be handled the same. This makes sure that the lint also covers `ptr::raw_const!(function)`, which the test should make sure actually works.